Improvements to attitude accuracy estimation when operating without external aiding

priseborough opened the pull request:
NOTE: the first patch is to enable reviewers to test in SITL and should not be pulled in as part of this PR.

Testing had shown that the legacy EKF outperformed both DCM and EKF2 for attitude accuracy when operating without GPS or optical flow aiding. This was traced to differences in the fusion methos used to constraint roll/pitch and yaw drift and tuning of the filter. The second patch addresses these issues with the following changes:

Instead of fusing in a constant velocity to constrain roll/pitch drift, it fuses in a constant position. This reduces the attitude error induced by rapid velocity changes during indoor flight. This fusion method was already available within the filter.

The more accurate 3-axis magnetometer fusion method can now be used when operating without aiding, whereas previously it was only available in combination with external aiding. Setting EK2_MAG_CAL =4 can now be used to force it to be used all the time, which may be be better for some users.

Learning of gyro scale factors is inhibited when aiding is not available. This was done because testing showed that this learning was not reliable in the absence of external aiding.

This patch has been tested on a Iris+ and bebop2 with GPS disabled. 

Here are graphs comparing roll and pitch outputs for the EKF (EKF1), EKF2 (NKF1) and DCM (AHR2) solutions from the bebop2 test with EK2_MAG_CAL=4. EKF2 and EKF are similar, however DCM does not return to level immediately after recovery from a manouevre. The noticeably degraded DCM behaviour is not evident in the Iris+ logs, indicating that there may be some board specific data processing issues to be resolved. 
![pitch](https://cloud.githubusercontent.com/assets/3596952/12529536/b08c46d8-c20e-11e5-8b27-cbb9429ce297.png)
![roll](https://cloud.githubusercontent.com/assets/3596952/12529537/b0e85b08-c20e-11e5-9602-e5711e5a671c.png)
Height innovations are lower than for the old EKF so alt hold stability should be improved over the baseline.
![height innovations](https://cloud.githubusercontent.com/assets/3596952/12529590/dc357924-c210-11e5-883d-98114ea2133d.png)
The normalised innovations were generally low (see below)
![normalised innovations](https://cloud.githubusercontent.com/assets/3596952/12529549/65401c3a-c20f-11e5-869e-44a7fde0a889.png)
however inspection of the magnetometer innovations provides evidence of a time alignment error between the magnetometer and imu data within the filter. The filer assumes the mag data is delayed 60msec relative to the imu data, a value that was determined from testing on a pixhawk board and is probably not optimal for the bebop hardware.
![magnetometer innovations](https://cloud.githubusercontent.com/assets/3596952/12529569/eb85d0f0-c20f-11e5-9087-081d10666acf.png)
Compare this to a flight without GPS from an Iris+ fitted with a Pixhawk
![iris magnetometer innovations](https://cloud.githubusercontent.com/assets/3596952/12529584/5d385254-c210-11e5-86e0-866de1eaa3fb.png)

OXINARF commented on libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp:
```diff
@@ -168,18 +168,21 @@ void NavEKF2_core::SelectVelPosFusion()
+        gpsDataDelayed.pos.y = lastKnownPositionNE.y;
+        // when in flight only fuse synthetic measurements when rate of change of velocity is less than 1.1g
+        // to reduce attitude errors due to launch acceleration
+        if (accNavMag < 1.1f * GRAVITY_MSS || motorsArmed) {
```
Maybe I'm not understanding the comment correctly, but what I get from it is that it should only enter the if when it is in flight **and** the rate of change is less than 1.1g. From what is there now if the motors are armed it always gets in.

priseborough replied:
Yes, thanks for picking that up. That was intended to be 
if (accNavMag < 1.1f * GRAVITY_MSS || !motorsArmed) {

tridge commented:
Hi Paul, regarding the test script, I think we should create a Tools/scripts/tests directory and add the adjusted test in there, plus move the original flip test in there too. Building up a set of scripted SITL tests would be useful. As usual the hard bit will be finding a good name for each test :)
I can do that when I bring in the PR if that suits you.

rmackay9 commented:
This is a pretty important change affecting at least the Bebop2 so I've merged the bottom two commits.

jschall commented:
@tridge @rmackay9 @priseborough just wondering what's up with this? It is old and looks like it didn't get any attention.

rmackay9 commented:
@jschall, ah.  I pushed in the previous two commits.  It was just the unit test that didn't go in.

It looks like since then @priseborough has done a forced push since then.
We probably should have closed the original PR and Paul should have created a new one?

OXINARF commented:
@rmackay9 are you sure you pushed those two commits? I can't find them and I remember that, at the time, I read your comment and couldn't find them either, but then I forgot to write it here.

priseborough commented:
Yes, I saw that the PR had not been closed and the code in master did not have the fixes, so because the bugs in the heading fusion were contributors to angle errors when operating without aiding (the primary use case of this fusion method), I rebased this PR on current master and added the fixes.

rmackay9 commented:
merged, thanks!
